[GeoMechanicsApplication] Add multi-stage tests for the new line interface elements

avdg81 opened the pull request:
**📝 Description**

Added two Python tests that cover the behavior of the new line interface element in a two-stage scenario. The first test uses Dirichlet boundary conditions, whereas the second one uses Neumann boundary conditions. Both tests carry out two time steps per stage.

Note that this PR needs to wait for [another PR](https://github.com/KratosMultiphysics/Kratos/pull/12531) to be merged first, since that PR contains a fix that has been duplicated here. Once it has become available on `master`, we can remove it from this PR.

avdg81 committed:
Added a rudimentary integration test

So far, the test comprises a single stage with a single time step. It doesn't include any initial relative displacement nor any initial traction.

avdg81 committed:
Improved indentation of json file

avdg81 committed:
Extended the integration test to carry out two time steps

The prescribed displacements are now interpolated from two tables (one for each direction). The expected values have been updated, too.

avdg81 committed:
Some preparations before adding a second stage

avdg81 committed:
Added a second stage

avdg81 committed:
Added a new test using a Neumann boundary condition

So far, this test consists of a single stage with two time steps.

avdg81 committed:
Added the second stage

avdg81 committed:
Both tests now use the "is restarted" flag

To make it work I had to copy a fix from another PR. Once that one has been merged, we can revert the local copy here.

avdg81 committed:
Minor refactoring work

- Renamed a test case to avoid having non-unique names.
- The interface normal stiffness and the interface shear stiffness are now attributes of the test case class.

avdg81 committed:
Compacted the test code by extracting a helper method

avdg81 committed:
Compacted the code for the test with Neumann BCs

WPK4FEM commented on applications/GeoMechanicsApplication/tests/line_interface_elements/Dirichlet_multi_stage/MaterialParameters.json:
```diff
@@ -0,0 +1,16 @@
+{
```
The Neumann and the Dirichlet version of the test are designed to give comparable results. That requires that the material is identical, which here could be achieved by using the same file for the mateial definition. That file is available in directory common. I suggest to use that one.

avdg81 replied:
Yes, that was exactly the intention, but it hadn't been implemented yet, since the branch was lagging behind a bit when you reviewed it. Your suggestion has now been processed. Thank you for reminding me.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_helper.py:
```diff
@@ -277,6 +284,32 @@ def get_cauchy_stress_tensor(simulation):
+    :param simulation:
+    :return: cauchy stress vector
+    """
+    model_part = simulation._list_of_output_processes[0].model_part
```
The 0 is very specific the first output process. That may contain a model part that is not able to produce CAUCHY_STRESS_VECTOR or on line 309 strain

avdg81 replied:
This helper function had been copied from another (related) PR with the intention to reuse it when it had become generally available. Later it was decided to replace this helper function by calling another helper function. The test code in this PR now follows the same approach. In other words, this helper function has been removed here, too.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_line_interface_elements.py:
```diff
@@ -81,6 +82,68 @@ def assert_outputs_for_3_plus_3_line_interface_element(self, simulation):
             self.assertAlmostEqual(relative_displacement_vertical_element[index][1], expected_vertical_displacement)
 
 
+    def assertVectorsAlmostEqual(self, actual_vectors, expected_vectors):
```
Isn't this asserVectorAlmostEqual to general to be in the script for a specific test. I would expect this in a helper script.

avdg81 replied:
Good catch. It is more general indeed. However, it can't be moved to, e.g. `test_helper.py`, since its implementation calls another assertion method, which is only known to test case instances. In other words, this checker must be part of the test case class. Perhaps at some point in time it can be moved to the Kratos Core...?

I did see room for one more improvement: I have found that class `TestCase` has a method `assertVectorAlmostEqual`. By reusing this method, I was able to get rid of the inner loop.

WPK4FEM commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_solver.py:
```diff
@@ -205,6 +207,15 @@ def PrepareModelPart(self):
                                                   self.settings["nodal_smoothing"].GetBool())
 
         self.main_model_part.ProcessInfo.SetValue(KratosMultiphysics.STEP, 0)
+        self.computing_model_part_name = "porous_computational_model_part"
```
Unexpected for me that adding a test should lead to changes in our entry point geomechanics_solver.py
Please explain why to me at some point in time.

avdg81 replied:
As discussed, these changes were necessary since the multi-stage tests make use of the "restart" mechanism. However, that mechanism was broken for geomechanics analyses, and it was fixed by another PR that had not been merged at the time I wrote this code. The intention was to remove these changes as soon as that PR was merged. Now that it has, the duplicated changes have been removed. Thank you for asking me about them.

WPK4FEM reviewed (commented):
Fast first look.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/12693-add-multistage-tests-for-new-line-interface-elements

avdg81 committed:
Replaced two copies of `MaterialParameters.json` by referencing the common one

As a result, the name of one model part had to be changed ("Interface" => "Interfaces").

avdg81 committed:
Retrieving the traction vectors has been unified

As a result, the new helper function could be removed (it was copied from another PR).

avdg81 committed:
Retrieving the relative displacement vectors has been unified

As a result, the new helper function could be removed (it was copied from another PR).

avdg81 committed:
Reuse the method that checks two vectors for equality

Also removed an accidentally added blank line.

avdg81 committed:
Updated the test documentation

WPK4FEM commented on applications/GeoMechanicsApplication/tests/line_interface_elements/Dirichlet_multi_stage/ProjectParameters_stage1.json:
```diff
@@ -0,0 +1,147 @@
+            "scaling":         false
+        },
+        "problem_domain_sub_model_part_list": ["Interfaces"],
+        "processes_sub_model_part_list":      ["FixedNodes","AllNodes", "DisplacedNodes"],
```
Either no space after comma, or always a space after a comma. Preferably not mixed as it is here.

avdg81 replied:
Good catch! Fixed.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/line_interface_elements/Dirichlet_multi_stage/ProjectParameters_stage2.json:
```diff
@@ -0,0 +1,146 @@
+            "scaling":         false
+        },
+        "problem_domain_sub_model_part_list": ["Interfaces"],
+        "processes_sub_model_part_list":      ["FixedNodes","AllNodes", "DisplacedNodes"],
```
see comment about comma and space in previous projectparameters file.

avdg81 replied:
Done. Thank you.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/line_interface_elements/Neumann_multi_stage/ProjectParameters_stage1.json:
```diff
@@ -0,0 +1,159 @@
+            "scaling":         false
+        },
+        "problem_domain_sub_model_part_list": ["Interfaces"],
+        "processes_sub_model_part_list":      ["FixedNodes","AllNodes", "LoadedSide"],
```
also here comma and space.

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/line_interface_elements/Neumann_multi_stage/ProjectParameters_stage2.json:
```diff
@@ -0,0 +1,158 @@
+            "scaling":         false
+        },
+        "problem_domain_sub_model_part_list": ["Interfaces"],
+        "processes_sub_model_part_list":      ["FixedNodes","AllNodes", "LoadedSide"],
```
no space, space

avdg81 replied:
Done.

WPK4FEM reviewed (dismissed):
Minor typographical  comments left, nothing blocking.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_line_interface_elements.py:
```diff
@@ -11,8 +11,9 @@ class KratosGeoMechanicsInterfaceElementTests(KratosUnittest.TestCase):
-        # Code here will be placed BEFORE every test in this TestCase.
-        pass
+        self.model = Kratos.Model()
+        self.normal_stiffness = 3.0e7
+        self.shear_stiffness = 1.5e7
```
Would it be possible to re-use these in the single-stage, single-step tests?

avdg81 replied:
Yes, of course we can. Thanks, this makes the test code more consistent.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_line_interface_elements.py:
```diff
@@ -81,6 +82,71 @@ def assert_outputs_for_3_plus_3_line_interface_element(self, simulation):
+            self.assertVectorAlmostEqual(actual_vector, expected_vector)
+
+
+    def test_multi_stage_3_plus_3_line_interface_element_with_dirichlet_conditions(self):
```
Since these tests are also such that the prescribed displacements result in the same tractions as Neumann (and the other way around), we might be able to reduce duplication a bit (similar to the single-stage tests), since all the asserts can be the same on both tests

avdg81 replied:
Good catch! I have extracted a method that asserts the results of a multi-stage test.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_line_interface_elements.py:
```diff
@@ -81,6 +82,71 @@ def assert_outputs_for_3_plus_3_line_interface_element(self, simulation):
+        for file_name, displacement_vector in zip(project_parameters_file_names, prescribed_displacement_vectors):
+            stage = test_helper.make_geomechanics_analysis(self.model, os.path.join(file_path, file_name))
+
+            stage.Run()
```
Just for my understanding. Is the reason we can't use test_helper.run_stages because we need to do the asserts for each stage before running the next?

avdg81 replied:
Yes, that is exactly why I have set it up this way. The point is that we need to check results by looking at the database, since not all output items can be written to result files yet. When it its possible to write all needed output items to result files, we can simplify the test code further.

rfaasse commented on applications/GeoMechanicsApplication/tests/line_interface_elements/Dirichlet_multi_stage/ProjectParameters_stage1.json:
```diff
@@ -0,0 +1,147 @@
+            "verbosity":       0,
+            "tolerance":       1.0e-6,
+            "scaling":         false
+        },
```
In the end, I used this solver (a bit less of a hassle to set up and should do the same)
```suggestion
        "linear_solver_settings":             {
            "solver_type":     "LinearSolversApplication.sparse_lu"
        }
```

avdg81 replied:
Thanks for pointing this out. I have simplified the linear solver settings accordingly in the `ProjectParameters.json` files that have been added in this PR.

rfaasse commented on applications/GeoMechanicsApplication/tests/line_interface_elements/Dirichlet_multi_stage/ProjectParameters_stage1.json:
```diff
@@ -0,0 +1,147 @@
+        "buffer_size":                        2,
+        "echo_level":                         1,
+        "clear_storage":                      false,
+        "compute_reactions":                  true,
```
I think we could put this to false. I did some clean-up in the single stage tests (which is of course based of the same example), maybe you could diff and see which settings could be the same? The same holds then of course for the other jsons

avdg81 replied:
I have compared the single stage analysis files with those of the multi-stage analyses. For consistency, I have changed some flags for the multi-stage analyses which had no effect on the obtained results. Thanks for pointing this out.

rfaasse reviewed (commented):
Thanks for adding these tests (and nice you used the setup method for the test class), I have a few suggestions left for duplication/clean-up, but other than that, this is good to go!

avdg81 committed:
Removed a few redundant spaces

avdg81 committed:
Processed Richard's review comments

- Reuse attributes `normal_stiffness` and `shear_stiffness` in the single-stage tests.
- Extracted a method that asserts the results of a multi-stage test.
- Simplified the linear solver settings.
- Made the analysis flags more consistent with those of the single stage tests.

rfaasse reviewed (approved):
Looks good to go to me!
